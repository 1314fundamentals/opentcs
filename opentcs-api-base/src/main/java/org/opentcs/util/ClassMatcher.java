// SPDX-FileCopyrightText: The openTCS Authors
// SPDX-License-Identifier: MIT
package org.opentcs.util;

import java.io.Serializable;
import java.util.Arrays;
import java.util.HashSet;
import java.util.Set;
import java.util.function.Predicate;

/**
 * Checks whether the input argument is assignable to any of a given set of classes.
 */
public class ClassMatcher
    implements
      Predicate<Object>,
      Serializable {

  /**
   * The set of classes to evaluate incoming events to.
   */
  private final Set<Class<?>> clazzes = new HashSet<>();

  /**
   * Creates a new instance.
   *
   * @param events The set of classes to evaluate incoming events to.
   */
  public ClassMatcher(Class<?>... events) {
    clazzes.addAll(Arrays.asList(events));
  }

  @Override
  public boolean test(Object object) {
    return clazzes.stream().anyMatch(clazz -> clazz.isAssignableFrom(object.getClass()));
  }

  @Override
  public Predicate<Object> negate() {
    return new ClassMatcher() {
      @Override
      public boolean test(Object t) {
        return !ClassMatcher.this.test(t);
      }
    };
  }
}
