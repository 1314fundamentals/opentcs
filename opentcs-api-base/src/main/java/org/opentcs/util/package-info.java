// SPDX-FileCopyrightText: The openTCS Authors
// SPDX-License-Identifier: MIT
/**
 * A collection of classes providing miscellaneous functions that are used
 * throughout openTCS.
 */
package org.opentcs.util;
