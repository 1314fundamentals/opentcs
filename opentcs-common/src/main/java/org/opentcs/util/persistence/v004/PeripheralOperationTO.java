// SPDX-FileCopyrightText: The openTCS Authors
// SPDX-License-Identifier: MIT
package org.opentcs.util.persistence.v004;

import static java.util.Objects.requireNonNull;

import javax.xml.bind.annotation.XmlAccessType;
import javax.xml.bind.annotation.XmlAccessorType;
import javax.xml.bind.annotation.XmlAttribute;

/**
 */
@XmlAccessorType(XmlAccessType.PROPERTY)
public class PeripheralOperationTO
    extends
      PlantModelElementTO {

  private String locationName = "";
  private String executionTrigger = "";
  private boolean completionRequired;

  /**
   * Creates a new instance.
   */
  public PeripheralOperationTO() {
  }

  @XmlAttribute(required = true)
  public String getLocationName() {
    return locationName;
  }

  public PeripheralOperationTO setLocationName(String locationName) {
    this.locationName = requireNonNull(locationName, "locationName");
    return this;
  }

  @XmlAttribute(required = true)
  public String getExecutionTrigger() {
    return executionTrigger;
  }

  public PeripheralOperationTO setExecutionTrigger(String executionTrigger) {
    this.executionTrigger = requireNonNull(executionTrigger, "executionTrigger");
    return this;
  }

  @XmlAttribute(required = true)
  public boolean isCompletionRequired() {
    return completionRequired;
  }

  public PeripheralOperationTO setCompletionRequired(boolean completionRequired) {
    this.completionRequired = completionRequired;
    return this;
  }

}
