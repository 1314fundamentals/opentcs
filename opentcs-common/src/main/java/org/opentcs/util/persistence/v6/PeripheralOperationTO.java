// SPDX-FileCopyrightText: The openTCS Authors
// SPDX-License-Identifier: MIT
package org.opentcs.util.persistence.v6;

import static java.util.Objects.requireNonNull;

import javax.xml.bind.annotation.XmlAccessType;
import javax.xml.bind.annotation.XmlAccessorType;
import javax.xml.bind.annotation.XmlAttribute;
import javax.xml.bind.annotation.XmlType;

/**
 */
@XmlAccessorType(XmlAccessType.PROPERTY)
public class PeripheralOperationTO
    extends
      PlantModelElementTO {

  private String locationName = "";
  private ExecutionTrigger executionTrigger = ExecutionTrigger.AFTER_ALLOCATION;
  private boolean completionRequired;

  /**
   * Creates a new instance.
   */
  public PeripheralOperationTO() {
  }

  @XmlAttribute(required = true)
  public String getLocationName() {
    return locationName;
  }

  public PeripheralOperationTO setLocationName(String locationName) {
    this.locationName = requireNonNull(locationName, "locationName");
    return this;
  }

  @XmlAttribute(required = true)
  public ExecutionTrigger getExecutionTrigger() {
    return executionTrigger;
  }

  public PeripheralOperationTO setExecutionTrigger(ExecutionTrigger executionTrigger) {
    this.executionTrigger = requireNonNull(executionTrigger, "executionTrigger");
    return this;
  }

  @XmlAttribute(required = true)
  public boolean isCompletionRequired() {
    return completionRequired;
  }

  public PeripheralOperationTO setCompletionRequired(boolean completionRequired) {
    this.completionRequired = completionRequired;
    return this;
  }

  @XmlType
  public enum ExecutionTrigger {
    // CHECKSTYLE:OFF
    AFTER_ALLOCATION,
    AFTER_MOVEMENT;
    // CHECKSTYLE:ON
  }

}
