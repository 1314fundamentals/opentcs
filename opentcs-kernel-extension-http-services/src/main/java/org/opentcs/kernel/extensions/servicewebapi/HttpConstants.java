// SPDX-FileCopyrightText: The openTCS Authors
// SPDX-License-Identifier: MIT
package org.opentcs.kernel.extensions.servicewebapi;

/**
 * Defines some HTTP-related constants.
 */
public class HttpConstants {

  /**
   * Name of the header that is expected to contain the API access keys.
   */
  public static final String HEADER_NAME_ACCESS_KEY = "X-Api-Access-Key";
  /**
   * Content type for plain text.
   */
  public static final String CONTENT_TYPE_TEXT_PLAIN_UTF8 = "text/plain; charset=utf-8";
  /**
   * Content type for JSON structures.
   */
  public static final String CONTENT_TYPE_APPLICATION_JSON_UTF8 = "application/json; charset=utf-8";

  /**
   * Prevents instantiation.
   */
  private HttpConstants() {
  }
}
