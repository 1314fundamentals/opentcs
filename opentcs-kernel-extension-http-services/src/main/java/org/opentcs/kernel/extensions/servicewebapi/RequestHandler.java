// SPDX-FileCopyrightText: The openTCS Authors
// SPDX-License-Identifier: MIT
package org.opentcs.kernel.extensions.servicewebapi;

import io.javalin.apibuilder.EndpointGroup;
import org.opentcs.components.Lifecycle;

/**
 * A request handler.
 */
public interface RequestHandler
    extends
      Lifecycle {

  /**
   * Creates the handler's routes for the web service.
   */
  EndpointGroup createRoutes();
}
