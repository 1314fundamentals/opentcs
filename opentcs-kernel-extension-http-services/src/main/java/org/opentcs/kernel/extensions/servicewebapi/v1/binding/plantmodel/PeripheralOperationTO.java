// SPDX-FileCopyrightText: The openTCS Authors
// SPDX-License-Identifier: MIT
package org.opentcs.kernel.extensions.servicewebapi.v1.binding.plantmodel;

import static java.util.Objects.requireNonNull;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonProperty;
import jakarta.annotation.Nonnull;

/**
 */
public class PeripheralOperationTO {

  private String operation;
  private String locationName;
  private ExecutionTrigger executionTrigger = ExecutionTrigger.AFTER_ALLOCATION;
  private boolean completionRequired;

  @JsonCreator
  public PeripheralOperationTO(
      @Nonnull
      @JsonProperty(value = "operation", required = true)
      String operation,
      @Nonnull
      @JsonProperty(value = "locationName", required = true)
      String locationName
  ) {
    this.operation = requireNonNull(operation, "operation");
    this.locationName = requireNonNull(locationName, "locationName");
  }

  @Nonnull
  public String getOperation() {
    return operation;
  }

  public PeripheralOperationTO setOperation(
      @Nonnull
      String operation
  ) {
    this.operation = requireNonNull(operation, "operation");
    return this;
  }

  @Nonnull
  public String getLocationName() {
    return locationName;
  }

  public PeripheralOperationTO setLocationName(
      @Nonnull
      String locationName
  ) {
    this.locationName = requireNonNull(locationName, "locationName");
    return this;
  }

  @Nonnull
  public ExecutionTrigger getExecutionTrigger() {
    return executionTrigger;
  }

  public PeripheralOperationTO setExecutionTrigger(
      @Nonnull
      ExecutionTrigger executionTrigger
  ) {
    this.executionTrigger = requireNonNull(executionTrigger, "executionTrigger");
    return this;
  }

  public boolean isCompletionRequired() {
    return completionRequired;
  }

  public PeripheralOperationTO setCompletionRequired(boolean completionRequired) {
    this.completionRequired = completionRequired;
    return this;
  }

  // CHECKSTYLE:OFF
  public enum ExecutionTrigger {

    IMMEDIATE,
    AFTER_ALLOCATION,
    AFTER_MOVEMENT;
  }
  // CHECKSTYLE:ON
}
