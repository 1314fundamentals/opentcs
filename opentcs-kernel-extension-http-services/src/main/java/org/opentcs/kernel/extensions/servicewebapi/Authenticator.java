// SPDX-FileCopyrightText: The openTCS Authors
// SPDX-License-Identifier: MIT
package org.opentcs.kernel.extensions.servicewebapi;

import static java.util.Objects.requireNonNull;

import com.google.common.base.Strings;
import io.javalin.http.Context;
import jakarta.inject.Inject;
import java.util.Objects;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * Authenticates incoming requests.
 */
public class Authenticator {

  /**
   * This class's logger.
   */
  private static final Logger LOG = LoggerFactory.getLogger(Authenticator.class);
  /**
   * Defines the required access rules.
   */
  private final ServiceWebApiConfiguration configuration;

  /**
   * Creates a new instance.
   *
   * @param configuration Defines the required access rules.
   */
  @Inject
  public Authenticator(ServiceWebApiConfiguration configuration) {
    this.configuration = requireNonNull(configuration, "configuration");
  }

  /**
   * Checks whether authentication is required and the given request is authenticated.
   *
   * @param context The context to be checked.
   * @return <code>true</code> if, and only if, authentication is required and the given request is
   * authenticated.
   */
  public boolean isAuthenticated(Context context) {
    requireNonNull(context, "context");

    String requestAccessKey = context.header(HttpConstants.HEADER_NAME_ACCESS_KEY);
    LOG.debug(
        "Provided access key in header is '{}', required value is '{}'",
        requestAccessKey,
        configuration.accessKey()
    );

    // Any empty access key indicates authentication is not required.
    if (Strings.isNullOrEmpty(configuration.accessKey())) {
      LOG.debug("No access key, authentication not required.");
      return true;
    }

    return Objects.equals(requestAccessKey, configuration.accessKey());
  }

}
