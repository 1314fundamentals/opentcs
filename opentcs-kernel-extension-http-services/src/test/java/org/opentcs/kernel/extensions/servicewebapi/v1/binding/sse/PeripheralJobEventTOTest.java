// SPDX-FileCopyrightText: The openTCS Authors
// SPDX-License-Identifier: MIT
package org.opentcs.kernel.extensions.servicewebapi.v1.binding.sse;

import java.time.Instant;
import java.time.temporal.ChronoUnit;
import java.util.List;
import java.util.Map;
import java.util.TreeMap;
import org.approvaltests.Approvals;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.opentcs.kernel.extensions.servicewebapi.JsonBinder;
import org.opentcs.kernel.extensions.servicewebapi.v1.binding.sse.PeripheralJobEventTO.PeripheralJobTO.PeripheralOperationTO.ExecutionTriggerTO;
import org.opentcs.kernel.extensions.servicewebapi.v1.binding.sse.PeripheralJobEventTO.PeripheralJobTO.StateTO;

/**
 * Tests for {@link PeripheralJobEventTO}.
 */
class PeripheralJobEventTOTest {

  private JsonBinder jsonBinder;

  @BeforeEach
  void setUp() {
    jsonBinder = new JsonBinder();
  }

  @Test
  void jsonSample() {
    PeripheralJobEventTO to = new PeripheralJobEventTO(
        createPeripheralJob(StateTO.FINISHED),
        createPeripheralJob(StateTO.BEING_PROCESSED)
    );

    Approvals.verify(jsonBinder.toJson(to));
  }

  private PeripheralJobEventTO.PeripheralJobTO createPeripheralJob(StateTO state) {
    Instant time = Instant.parse("2025-07-22T12:03:00Z");
    return new PeripheralJobEventTO.PeripheralJobTO()
        .setName("job-1")
        .setProperties(new TreeMap<>(Map.of("key-1", "value-1", "key-2", "value-2")))
        .setHistory(
            new ObjectHistoryTO().setEntries(
                List.of(
                    new ObjectHistoryTO.ObjectHistoryEntryTO()
                        .setEventCode("event-code-1")
                        .setTimestamp(time)
                        .setSupplements(List.of("supplement-1", "supplement-2"))
                )
            )
        )
        .setReservationToken("reservation-token-1")
        .setRelatedVehicle("vehicle-1")
        .setRelatedTransportOrder("transport-order-1")
        .setPeripheralOperation(
            new PeripheralJobEventTO.PeripheralJobTO.PeripheralOperationTO()
                .setLocation("location-1")
                .setOperation("operation-1")
                .setExecutionTrigger(ExecutionTriggerTO.AFTER_MOVEMENT)
                .setCompletionRequired(true)
        )
        .setState(state)
        .setCreationTime(time)
        .setFinishedTime(time.plus(1, ChronoUnit.HOURS));
  }
}
