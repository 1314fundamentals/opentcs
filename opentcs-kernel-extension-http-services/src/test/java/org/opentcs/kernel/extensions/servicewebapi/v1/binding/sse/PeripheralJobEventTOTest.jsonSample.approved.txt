{
  "currentObjectState" : {
    "name" : "job-1",
    "properties" : {
      "key-1" : "value-1",
      "key-2" : "value-2"
    },
    "history" : {
      "entries" : [ {
        "timestamp" : "2025-07-22T12:03:00Z",
        "eventCode" : "event-code-1",
        "supplements" : [ "supplement-1", "supplement-2" ]
      } ]
    },
    "reservationToken" : "reservation-token-1",
    "relatedVehicle" : "vehicle-1",
    "relatedTransportOrder" : "transport-order-1",
    "peripheralOperation" : {
      "location" : "location-1",
      "operation" : "operation-1",
      "executionTrigger" : "AFTER_MOVEMENT",
      "completionRequired" : true
    },
    "state" : "FINISHED",
    "creationTime" : "2025-07-22T12:03:00Z",
    "finishedTime" : "2025-07-22T13:03:00Z"
  },
  "previousObjectState" : {
    "name" : "job-1",
    "properties" : {
      "key-1" : "value-1",
      "key-2" : "value-2"
    },
    "history" : {
      "entries" : [ {
        "timestamp" : "2025-07-22T12:03:00Z",
        "eventCode" : "event-code-1",
        "supplements" : [ "supplement-1", "supplement-2" ]
      } ]
    },
    "reservationToken" : "reservation-token-1",
    "relatedVehicle" : "vehicle-1",
    "relatedTransportOrder" : "transport-order-1",
    "peripheralOperation" : {
      "location" : "location-1",
      "operation" : "operation-1",
      "executionTrigger" : "AFTER_MOVEMENT",
      "completionRequired" : true
    },
    "state" : "BEING_PROCESSED",
    "creationTime" : "2025-07-22T12:03:00Z",
    "finishedTime" : "2025-07-22T13:03:00Z"
  }
}